Issue 6

nicholas-esterer opened the pull request:
Initial draft of ML image segmentation web app.

emmanuelle commented on dash_utils.py:
```diff
@@ -0,0 +1,18 @@
+import dash
```
Could you please explain how this file is used?

nicholas-esterer replied:
I've removed it, it is no longer necessary

emmanuelle commented on update_stroke_colour.py:
```diff
@@ -0,0 +1,114 @@
+import plotly.express as px
```
is this a different version of the app?

nicholas-esterer replied:
I accidentally included this file, which was a test. I've deleted this, too.

emmanuelle commented on assets/test_masks/0.json:
```diff
@@ -0,0 +1 @@
+{"editable": true, "layer": "above", "line": {"color": "LightSeaGreen", "dash": "solid", "width": 10}, "opacity": 1, "path": "M40.53888888888886,32.63333333333333L37.572222222222194,32.63333333333333L14.827777777777749,22.744444444444444L4.938888888888861,10.877777777777778L0.9833333333333055,2.966666666666667L0.9833333333333055,0L15.816666666666638,0.9888888888888888L33.61666666666664,11.8666666 ...
```
why do you need the json and png binary files?

nicholas-esterer replied:
we don't, they're left over from testing which I forgot to remove...

emmanuelle commented on ml_image_segmentation.py:
```diff
@@ -0,0 +1,224 @@
+# the number of different classes for labels
+NUM_LABEL_CLASSES = 15
+DEFAULT_LABEL_CLASS = 0
+class_label_colormap = px.colors.qualitative.Light24
```
would it be possible to use the same colors for the shapes and for the segmented regions?

nicholas-esterer replied:
this would be nice, it's true, let me take a look

emmanuelle commented on ml_image_segmentation.py:
```diff
@@ -0,0 +1,224 @@
+            options=[
+                {"label": l.capitalize(), "value": l } for l in SEG_FEATURE_TYPES
+            ],
+            value=[]
```
could you please initialize with intensity and edges? If the value is empty nothing will happen for the segmentation (it happened to me as I was testing the app)

nicholas-esterer replied:
done :+1:

emmanuelle commented on update_stroke_colour.py:
```diff
@@ -0,0 +1,114 @@
+import utils
+import shape_utils
+import plot_common
+import flask
```
I think flask is not used.

nicholas-esterer replied:
this file also not used

emmanuelle commented on requirements.txt:
```diff
@@ -1,9 +1,131 @@
+appdirs==1.4.3
```
was there something else than `cairosvg` which needed to be added here? `dash_canvas` is not used since you used a plotly validator for the image string. Probably you can just add `cairosvg` manually to the version in master.

nicholas-esterer replied:
I'm not sure, but is there a way to automatically determine what needs to be included in the `requirements.txt` file?

emmanuelle replied:
https://pypi.org/project/pipreqs/ looks quite nice for this

emmanuelle replied:
although I believe the number of deps is still small enough to track them manually ;-)

nicholas-esterer replied:
`pipreqs` works great :+1:

nicholas-esterer replied:
ah but it's true it forgot some stuff which isn't imported in any of the scripts, e.g., `gunicorn`

emmanuelle commented on shape_utils.py:
```diff
@@ -0,0 +1,89 @@
+import numpy as np
+
+
+def shape_to_svg_code(fig=None, shape=None, width=None, height=None):
```
if `shape` should not be None, please pass it as a first positional argument and remove the assert about shape.

nicholas-esterer replied:
fixed :+1:

emmanuelle commented on shape_utils.py:
```diff
@@ -0,0 +1,89 @@
+        wrange = next(fig.select_xaxes())["range"]
+        hrange = next(fig.select_yaxes())["range"]
+        width, height = [max(r) - min(r) for r in [wrange, hrange]]
+    assert (width is not None) and (height is not None)
```
please raise an error rather than using assert (see for example https://medium.com/@jadhavmanoj/python-what-is-raise-and-assert-statement-c3908697bc62)

nicholas-esterer replied:
:+1:

emmanuelle commented:
I left a couple of comments. Tomorrow I'll read some parts of the code in a closer way but if I add comments they will probably be only stylistic. Here is also a wishlist which does not need to be addressed in this PR but more for the roadmap of the app
- add a `dcc.RangeSlider` for the sigma min and max in `trainable_segmentation`
- use a `dcc.Loading` component when the callback runs so that users know they should wait a few seconds https://dash.plotly.com/dash-core-components/loading 
- if we don't want the segmentation to run again every time a segmentation feature is changed, we could use `segmentation-features` as a State rather than an Input and trigger the segmentation only when "show segmentation" is the callback context and is checked. I'm not completely sure about the best UI here.

emmanuelle commented:
I saw you removed the css files from the bounding-box app, was it because you didn't want to use them for the machine learning app? At some point I think we need to use a similar structure as in https://github.com/plotly/dash-sample-apps/ with one directory per app. If you clone this repo be sure to do it when you have time it's very big :-).

nicholas-esterer commented:
I removed the CSS because I think we had to do some weird stuff to make the layout for the bounding box app work. When we split the apps up we can put back some CSS. Or I can add paths to ignore to `dash.Dash()`.

emmanuelle commented:
> I removed the CSS because I think we had to do some weird stuff to make the layout for the bounding box app work. When we split the apps up we can put back some CSS. Or I can add paths to ignore to `dash.Dash()`.

yes we would really need one `assets` directory per app, so let's wait until we split the apps.

emmanuelle commented on requirements.txt:
```diff
@@ -1,9 +1,13 @@
+dash_table==4.7.0
+Pillow==7.1.2
+scikit_learn==0.23.1
+skimage==0.0
```
there's something weird with the version of scikit-image I think?

nicholas-esterer replied:
I fixed this, plus it was missing `pandas` and `gunicorn`. `pipreq` can't find stuff not imported I guess (how would it haha) :man_shrugging:

emmanuelle commented:
Just a tiny comment about the `requirements.txt`, so :woman_dancing: !
